src/FsAutoComplete/LspServers/AdaptiveServerState.fs

```diff
@@ -2051,26 +2129,45 @@ type AdaptiveState(lspClient: FSharpLspClient, sourceTextFactory: ISourceTextFac
             && file.Contains "AssemblyAttributes.fs" |> not)
 
         let checksToPerformLength = innerChecks.Length
+        let rootToken = sourceFilePath |> getOpenFileTokenOrDefault
 
         innerChecks
-        |> Array.map (fun (proj, file) ->
-          let file = UMX.tag file
+        |> Array.map (fun (opts, file) ->
+          let fileTagged = Utils.normalizePath file
+          async {
+
+            use joinedToken =
+              if fileTagged = sourceFilePath then
+                // dont reset the token for the incoming file as it would cancel the whole operation
+                CancellationTokenSource.CreateLinkedTokenSource(rootToken, progressReporter.CancellationToken)
+              else
+                // only cancel other files
+                // If we have multiple saves from separate root files we want only one to be running
+                let token = resetCancellationToken fileTagged None
+                // and join with the root token as well since we want to cancel the whole operation if the root files changes
+                CancellationTokenSource.CreateLinkedTokenSource(rootToken, token, progressReporter.CancellationToken)
```

Comment:
Saves now will try to cancel other potential typecheck operations for the same file. And clicking cancel will stop the overall operation.

Technically we're really want a graph directed version of cancellation tokens but this will be as good as we get without FCS letting us use the graph.